includes/core-changes/codeanalysis/5.0/ca1416-platform-compatibility-analyzer.md

```diff
@@ -0,0 +1,125 @@
+
+Rule CA1416, the platform compatibility analyzer, works hand-in-hand with some other features that are new in .NET 5.0. .NET 5.0 introduces the <xref:System.Runtime.Versioning.MinimumOSPlatformAttribute> (soon to be `SupportedOSPlatformAttribute`) and <xref:System.Runtime.Versioning.RemovedInOSPlatformAttribute> (soon to be `UnsupportedOSPlatformAttribute`) attributes, which let you specify which platforms an API *is* or *isn't* supported on. In the absence of these attributes, an API is assumed to be supported on all platforms. These attributes have been applied to platform-specific APIs in the core .NET libraries.
+
+In projects that target platforms for which APIs that they use aren't available, rule CA1416 flags any platform-specific API call where the platform context isn't verified. Most of the APIs that are now decorated with the `SupportedOSPlatformAttribute` and `UnsupportedOSPlatformAttribute` attributes throw <xref:System.PlatformNotSupportedException> exceptions when they're invoked on an unsupported operating system. Now that these APIs are marked as platform-specific, rule CA1416 helps you prevent run-time <xref:System.PlatformNotSupportedException> exceptions by adding OS checks to your call sites.
+
+#### Examples
+
+- The <xref:System.Console.Beep(System.Int32,System.Int32)?displayProperty=nameWithType> method is only supported on Windows (it's decorated with `[SupportedOSPlatform("windows")]`). The following code will produce a CA1416 warning at build time if the project [targets](../../../../docs/standard/frameworks.md) `net5.0` (but not `net5.0-windows`). For actions you can take to avoid the warning, see [Recommended action](#recommended-action).
+
+  ```csharp
+  public void PlayCMajor()
+  {
+      Console.Beep(261, 1000);
+  }
+  ```
+
+- The <xref:System.Drawing.Image.FromFile(System.String)?displayProperty=nameWithType> method is not supported in the browser (it's decorated with `[UnsupportedOSPlatform("browser")]`). The following code will produce a CA1416 warning at build time if the project uses the Blazor WebAssembly SDK (`<Project Sdk="Microsoft.NET.Sdk.BlazorWebAssembly">`) or includes `browser` as a supported platform (`<SupportedPlatform Include="browser" />`) in the project file.
+
+  ```csharp
+  public void CreateImage()
+  {
+      Image newImage = Image.FromFile("SampImag.jpg");
+  }
+  ```
+
+#### Version introduced
+
+5.0 RC1
+
+#### Recommended action
+
+Ensure that platform-specific APIs are only called when the code is running on an appropriate platform. You can achieve this using preprocessor directives, or at run time by using platform guards:
+
+- Add a `#if` [preprocessor directive](../../../../docs/csharp/language-reference/preprocessor-directives/preprocessor-if.md) around platform-specific API calls:
```

Comment:
I believe the user would also need to _define_ this constant too, with something like the following.  Is that right, @dsplaisted?

```xml
<PropertyGroup>
	<DefineConstants Condition="'$(TargetFramework)' == 'net5.0-windows'">NET50_WINDOWS</DefineConstants>
</PropertyGroup>
```

Reply:
[This section](https://docs.microsoft.com/en-us/dotnet/standard/frameworks#how-to-specify-a-target-framework) says they're predefined, but I should add the .NET 5 symbols to the table (last part of the section).

Reply:
Ah, cool; thanks. I'm not seeing this work on my machine, but I'm not sure why.

Reply:
@jeffhandley Can you try NET5_0-WINDOWS? I think I need to correct this snippet.

Reply:
I've tried TFMs of `net5.0-windows`, `net5.0-windows7.0`, and `net5.0-windows10.0.19041`, and with each of them `DefineConstants` gets produced as `TRACE;DEBUG;NET;NET5_0`.

When trying to use `#if NET5_0-WINDOWS`, I also get an error that the `-` is an unexpected character.

Reply:
Probably the hyphen needs to be changed to an underscore too, so `NET5_0_WINDOWS`. But it sounds like that still wouldn't be valid...

Reply:
Yeah, I tried that before checking what constants are getting generated.  I wonder if we should just remove that section until we get this detail worked out.

Reply:
@gewarren Based on a meeting earlier this week related to what constants will get defined, we are not yet defining the platform-specific constants like this in the SDK, so I think we're good to leave this `#if def` section out permanently.

Reply:
Thanks for updating me.